runtime/serde/serde-xml/common/src/aws/smithy/kotlin/runtime/serde/xml/tokenization/XmlScanner.kt

```diff
@@ -0,0 +1,368 @@
+/*
+ * Copyright Amazon.com, Inc. or its affiliates. All Rights Reserved.
+ * SPDX-License-Identifier: Apache-2.0.
+ */
+package aws.smithy.kotlin.runtime.serde.xml.tokenization
+
+import aws.smithy.kotlin.runtime.serde.DeserializationException
+import aws.smithy.kotlin.runtime.serde.xml.XmlToken
+import aws.smithy.kotlin.runtime.util.text.codePointToChars
+
+// https://www.w3.org/TR/xml/#NT-S
+private val xmlWhitespaceChars = setOf(
+    '\t', // horizontal tab
+    '\r', // carriage return
+    '\n', // newline
+    ' ', // normal space
+)
+private fun Char.isXmlWhitespace() = xmlWhitespaceChars.contains(this)
+
+private fun Char.toRange() = this..this
+
+// https://www.w3.org/TR/xml/#NT-Name
+private val nameStartCharRanges = setOf(
+    ':'.toRange(),
+    'A'..'Z',
+    '_'.toRange(),
+    'a'..'z',
+    '\u00C0'..'\u00D6',
+    '\u00D8'..'\u00F6',
+    '\u00F8'..'\u02FF',
+    '\u0370'..'\u037D',
+    '\u037F'..'\u1FFF',
+    '\u200C'..'\u200D',
+    '\u2070'..'\u218F',
+    '\u2C00'..'\u2FEF',
+    '\u3001'..'\uD7FF',
+)
+private val nameCharRanges = nameStartCharRanges + setOf(
+    '-'.toRange(),
+    '.'.toRange(),
+    '0'..'9',
+    '\u00B7'.toRange(),
+    '\u0300'..'\u036f',
+    '\u203f'..'\u2040',
+)
+
+// TODO these probably aren't very performant. We may be able to improve efficiency with a treeset of ranges
+private fun Char.isValidForNameStart(): Boolean = nameStartCharRanges.any { it.contains(this) }
+private fun Char.isValidForName(): Boolean = nameCharRanges.any { it.contains(this) }
+
+private val decimalCharRef = "#([0-9]+)".toRegex()
+private val hexCharRef = "#x([0-9a-fA-F]+)".toRegex()
+
+// https://www.w3.org/TR/xml/#sec-predefined-ent
+private val namedRefs = mapOf(
+    "lt" to '<',
+    "gt" to '>',
+    "amp" to '&',
+    "apos" to '\'',
+    "quot" to '"',
+).mapValues { charArrayOf(it.value) }
+
+private fun XmlToken.QualifiedName.isXmlns(): Boolean = (local == "xmlns" && prefix == null) || (prefix == "xmlns")
+private fun XmlToken.QualifiedName.xmlnsPrefix(): String? = if (local == "xmlns") null else local
+
+private typealias AttributeMap = Map<XmlToken.QualifiedName, String>
+private fun AttributeMap.extractNsDeclarations(): Pair<AttributeMap, List<XmlToken.Namespace>> {
+    val parts = toList().partition { it.first.isXmlns() }
+    val attributes = parts.second.toMap()
+    val nsDeclarations = parts.first.map { XmlToken.Namespace(it.second, it.first.xmlnsPrefix()) }
+    return attributes to nsDeclarations
+}
+
+/**
+ * A stateful scanner that scans a [StringTextStream] and reads [XmlToken] elements.
+ */
+class XmlScanner(internal val source: StringTextStream) {
```

Comment:
nit

This seems like it is should be the one named `XmlLexer` since it doing the actual lexical analysis

Reply:
You're right, originally this was all inside what's now called `XmlLexer`. Will rename _this_ class to `XmlLexer` and rename the other class to `LexingXmlStreamReader` (since it's primarily an adapter for the `XmlStreamReader` interface and uses a lexer as its primary delegate).